src/python/zquantum/core/wip/estimators/estimation.py

```diff
@@ -0,0 +1,233 @@
+    if number_of_shots <= 0:
+        raise ValueError("number_of_shots must be positive.")
+
+    def _allocate_shots(
+        estimation_tasks: List[EstimationTask],
+    ) -> List[EstimationTask]:
+
+        return [
+            EstimationTask(
+                operator=estimation_task.operator,
+                circuit=estimation_task.circuit,
+                number_of_shots=number_of_shots,
+            )
+            for estimation_task in estimation_tasks
+        ]
+
+    return _allocate_shots
+
+
+def proportional_shot_allocation(
+    total_n_shots: int,
+    prior_expectation_values: Optional[ExpectationValues] = None,
+) -> EstimationTaskTransformer:
+    """
+    Returns an EstimationTaskTransformer which allocates the same number of shots to each task.
+    For more details please refer to documentation of zquantum.core.hamiltonian.estimate_nmeas_for_frames .
+
+    Args:
+        total_n_shots: total number of shots to be allocated
+        prior_expectation_values: object containing the expectation
+            values of all operators in frame_operators
+    """
+    if total_n_shots <= 0:
+        raise ValueError("total_n_shots must be positive.")
+
+    def _allocate_shots(
+        estimation_tasks: List[EstimationTask],
+    ) -> List[EstimationTask]:
+        frame_operators = [
+            estimation_task.operator for estimation_task in estimation_tasks
+        ]
+
+        _, _, measurements_per_frame = estimate_nmeas_for_frames(
+            frame_operators, prior_expectation_values
+        )
+
+        measurements_per_frame = scale_and_discretize(
+            measurements_per_frame, total_n_shots
+        )
+
+        return [
+            EstimationTask(
+                operator=estimation_task.operator,
+                circuit=estimation_task.circuit,
+                number_of_shots=number_of_shots,
+            )
+            for estimation_task, number_of_shots in zip(
+                estimation_tasks, measurements_per_frame
+            )
+        ]
+
+    return _allocate_shots
+
+
+def naively_estimate_expectation_values(
+    backend: QuantumBackend,
+    estimation_tasks: List[EstimationTask],
+) -> ExpectationValues:
+    """
+    Basic method for estimating expectation values for list of estimation tasks.
+    It executes specified circuit and calculates expectation values based on the measurements.
+
+    Args:
+        backend: backend used for executing circuits
+        estimation_tasks: list of estimation tasks
+    """
+    circuits, operators, shots_per_circuit = zip(
+        *[(e.circuit, e.operator, e.number_of_shots) for e in estimation_tasks]
+    )
+
+    measurements_list = backend.run_circuitset_and_measure(circuits, shots_per_circuit)
+
+    expectation_values_list = [
+        expectation_values_to_real(measurements.get_expectation_values(frame_operator))
+        for frame_operator, measurements in zip(operators, measurements_list)
+    ]
+
+    # TODO handle empty term?
```

Comment:
The reason this logic was included in the `BasicEstimator` was that the grouping functions [throw away the constant term](https://github.com/zapatacomputing/z-quantum-core/blob/943f28bb23f404ad1f1c7c2f68ce3656993e692d/src/python/zquantum/core/hamiltonian.py#L58-L59). I think that, as currently written, `greedy_grouping_with_context_selection` + `naively_estimate_expectation_values` will yield the wrong answer because the constant term will get lost.

One way to address this could be to just implement [the story about making the constant term included everywhere](https://www.pivotaltracker.com/story/show/176774633). I think the needed changes are:
* Change `group_comeasureable_terms_greedy` to keep the constant term
* Update [the check for the constant term in `compute_group_variances`](https://github.com/zapatacomputing/z-quantum-core/blob/943f28bb23f404ad1f1c7c2f68ce3656993e692d/src/python/zquantum/core/hamiltonian.py#L122-L125
)
* Remove [the logic to exclude the constant term](https://github.com/zapatacomputing/z-quantum-core/blob/d2da45b051d0460cffcd6351470e23c0458fa3dd/src/python/zquantum/core/hamiltonian.py#L206-L209) from `get_expectation_values_from_rdms`.

What do you think about trying to implement the above changes in a separate PR?

Reply:
Sounds good, will make a PR to this branch.